CIP-???? | Efficient Proofs for Dynamic Sets (#816)
cardano-foundation/CIPs: CIP-????/README.md

---
CIP: XXXX
Title: Efficient Proofs for Dynamic Sets on Cardano
Status: Proposed
Category: Tools
Authors:
    - Pal Dorogi <pal.dorogi@gmail.com>

> Discussions:

Reviewer: ```suggestion
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/816
```

The use of IntegriTree as the off-chain data structure allows for efficient and secure tracking of the minting and burning of NFTs. IntegriTree is a new accumulator scheme that leverages a complete binary hash tree for compact accumulation and dynamic updates. 

The tree is used to accumulate the elements in the set, and to provide proofs. 
The values are represented as open-interval pairs of consecutive elements with each node in the tree. IntegriTree offers proofs for membership, non-membership, addition, and deletion, with a proof size that scales logarithmically with the number of elements.

The use of an accumulator to represent the state of the IntegriTree allows for the updated state to be verified and recorded on the blockchain in a single transaction. This simplifies the process of updating the state, and reduces the risk of errors or disputes.

## Path to Active

### Acceptance Criteria
The acceptance criteria for this CIP to become active are as follows:

- The solution described in the CIP has been implemented and tested on the Cardano blockchain.

> - The implementation has been reviewed and approved by the Cardano Foundation and IOHK.

Reviewer 2: ```suggestion
- The implementation has been reviewed and approved by subject matter experts.
```
Same rationale as the comment below r.e. not requiring a participant from a particular organization but instead someone whose personal credentials make them an expert worthy of providing peer review to the subject at hand.

> - The solution described in the CIP has been implemented and tested on the Cardano blockchain.
> - The implementation has been reviewed and approved by the Cardano Foundation and IOHK.
> - The community has had sufficient time to review and provide feedback on the CIP and the implementation.
> - Any concerns or issues raised during the review and testing process have been addressed and resolved.

Reviewer 2: ```suggestion
[ ] The solution described in the CIP has been implemented and tested on the Cardano blockchain.
[ ] The implementation has been reviewed and approved by the Cardano Foundation and IOHK.
[ ] The community has had sufficient time to review and provide feedback on the CIP and the implementation.
[ ] Any concerns or issues raised during the review and testing process have been addressed and resolved.
```
Change to checkboxes to match the formatting of other CIPs and allow for easy tracking of progress of the proposal.

## Implementation Plan

The implementation plan for this CIP is as follows:

- The solution described in the CIP will be implemented and tested - on the Cardano blockchain by the proposer of the CIP.

> - The proposer will submit the implementation for review and - approval by the Cardano Foundation and IOHK.

Reviewer 2: ```suggestion
- The proposer will submit the implementation for review and approval by subject matter experts.
```
While it would be great to have participants from both CF and IOHK/IOI/IOG... And we can certainly tag up some of those individuals. I feel it is best to not explicitly require participation of any particular entity and would rather judge the readiness of the proposal based on the reviewer's credentials instead.

> - The solution described in the CIP will be implemented and tested - on the Cardano blockchain by the proposer of the CIP.
> - The proposer will submit the implementation for review and - approval by the Cardano Foundation and IOHK.
> - The proposer will make the implementation and the CIP available for review and feedback by the community.
> - The proposer will address and resolve any concerns or issues raised during the review and testing process.
> - Once the acceptance criteria have been met, the proposer will submit a pull request to the Cardano Improvement Proposals repository to update the status of the CIP to active.

Reviewer 2: ```suggestion
[ ] The solution described in the CIP will be implemented and tested - on the Cardano blockchain by the proposer of the CIP.
[ ] The proposer will submit the implementation for review and - approval by the Cardano Foundation and IOHK.
[ ] The proposer will make the implementation and the CIP available for review and feedback by the community.
[ ] The proposer will address and resolve any concerns or issues raised during the review and testing process.
[ ] Once the acceptance criteria have been met, the proposer will submit a pull request to the Cardano Improvement Proposals repository to update the status of the CIP to active.
```
Change to checkboxes to match the formatting of other CIPs and allow for easy tracking of progress of the proposal.

> ## Optional sections
> 
> ### Versioning

Reviewer: ```suggestion
## Versioning
```
The _Versioning_ section is much appreciated, although "Optional sections" is not actually a section 😅

---
CIP: XXXX
Title: Efficient Proofs for Dynamic Sets on Cardano
Status: Proposed
Category: Tools
Authors:
    - Pal Dorogi <pal.dorogi@gmail.com>
Implementors: []
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/816
    - https://github.com/cardano-foundation/CIPs/pull/816
Created: 2024-05-17
License: CC-BY-4.0
---

## Abstract

This Cardano Improvement Proposal (CIP) describes a solution for ensuring the uniqueness (at minting time) of dynamically minted NFTs on the Cardano blockchain. 

> The solution uses a new accumulator scheme, leverages a complete binary hash tree for compact accumulation and dynamic updates which relies on a trusted Plutus script, specifically designed to perform accumulator verification.

Author: Yeah, it's ambiguous and misleading. It simply means, that the deployed Plutus script is secure in the sense that no any adversarial can modify/change it after deployment, but it does not mean that Plutus script itself is secure (it requires some audit etc.) in any way.

> The solution uses a new accumulator scheme, leverages a complete binary hash tree for compact accumulation and dynamic updates which relies on a trusted Plutus script, specifically designed to perform accumulator verification.

Reviewer 3: Ah, is this the script that everyone will use? In that, you aim to create a universal script that others can reference, so that you can onchain proof that projects adhere to this standard.

When a user wants to mint or burn an NFT, they rebuild the tree from the blockchain (or can use any 3rd-party service provider), 
create a new root hash or accumulator for the modified (added/deleted element) tree, and a proof, and send these data to the blockchain. 

The on-chain code verifies the proof and the updated state, and if the verification is successful, the minting or burning transaction is processed.

## Motivation: why is this CIP necessary?

NFTs are unique digital assets that can represent ownership of a wide range of items, such as art, collectibles, and in-game items. One challenge with minting NFTs on the Cardano blockchain is ensuring the uniqueness of each token at minting time. When a user wants to mint an NFT using a decentralized application (dApp), the on-chain code is unable to verify that the NFT being minted with the same policy ID is unique. 

> This can lead to the creation of duplicate NFTs, which can cause confusion and disputes over ownership.

Author: Yes, it does not assume any knowledge 

> So I think this section or some other should specify more that the problem you are addressing is when assets are not completely know yet before deployment (correct me if my assessment is wrong).

Yes, you're correct, this solution does not assume any knowledge of the mint-able assets.

  - The position of a node in the tree is determined by **`P = i - 2^d + 1`** which means it's the **`Pth`** node on level `d` of the `T` tree.
  - The **left**/**right** position within a node follows the **`even-odd`** rule:  if its index `i` is even, the child is on the left side; otherwise, it's on the right. This position is determined by **`Pn = In - 2Ip`** (`Ip` is the parent's index)  where `Pn` is 0 when the child is on the left, 1 when it's on the right, and results in an **error** otherwise.
- In order to keep track of changes in the state of the `T` tree, there are two types of nodes:
  - Updateable node (**`Nu`**): These nodes are already part of the tree, but their values will be updated sometime by an update process.
  - Parent node (**`Np`**): These are then nodes where a new leaf `L` will be appended to or the last leaf removed from.
- The right child (`Nr`) of an parent node (`Np`), must always be an empty hash, represented as **`ε`**. The left child can be either an empty hash (**`ε`**) or a leaf.

## Rationale: how does this CIP achieve its goals?

> The solution described in this CIP achieves the goal of ensuring the uniqueness of NFTs minted on the Cardano blockchain by using an off-chain data structure to track the minting and burning of NFTs, and storing the hash of this data structure together with the proofs on the blockchain. This allows the on-chain code to verify the uniqueness of an NFT at minting time, without requiring the entire data structure to be stored on the blockchain.

Author: Yes, your assumption is correct. Though, that's why the title contains `dynamic set`, though the name can be misleading/confusing, as there is no proper definition what I meant by `dynamic set` (dynamically add/remove elements without any constraints).

---
CIP: XXXX

> Title: Efficient Proofs for Dynamic Sets on Cardano

Author: @rphair, thanks for that. I would say the CPS would be a better option, as there could be more than one solution to solve the problem. In fact, I have another proposal that is based on BLS12-381.

Typescript's example:
``` typescript

export type Val =  { xi: string; xa: string; xb: string }

export type IntegriTree = {
  val: Val;
  left: IntegriTree;
  right: IntegriTree;
}
```

> Note: the current typescript implementation of IntegriTree uses an array of Vals for efficient operations.

### Root Hash Calculation of IntegriTree

The off-chain root hash calculateion of the tree or any of its branch is calculated using the following function:
``` typescript

  /**
   * Calculates the hash for a specified node in the tree.
   * @param index - The index of the node.
   * @returns The hash for the specified node.
   */
  private hashNode(index: number): string {
    if (index >= this.elements.length) {
      return emptyHash
    }

    const val = this.elements[index]
    const leftChildIndex = 2 * index + 1
    const rightChildIndex = 2 * index + 2

    const valHash = hashVal(val)
    const leftHash = this.hashNode(leftChildIndex)
    const rightHash = this.hashNode(rightChildIndex)

    return combineThreeHashes(valHash, leftHash, rightHash)
  }
```

> Note: A compact array based structure is used for off-chain representation of the tree.

> ## Minimal Subtree (Proof)

Reviewer 4: ```suggestion
### Minimal Subtree (Proof)
```

In merkle tree the leaf pairs (that contain data) are hashed up to the tree while in IntegriTree the nodes' values (`Val`) are part of the proof.
For example, for the following Proof

``` typescript
const proof = HashNode {
  hash: valHash(Val{ xi: 1, xa: "a", xb: "adam"}),
  left: HashNode {
            hash: valHash(Val{ xi: 2, xa: "adam", xb: "b"}),
            left: emptyHash,
            right: emptyHash
        },
  right: emptyHash
}
```

It's easily can be proven that "adam" is in the tree, but "aby" is not, having a `Val { xi: "1", xa: "a", xb: "adam"}` and the Proof
as the proof's `rootHash(proof)` exactly the same with the rootHash of the whole integritree, and the hash of proof contains the hash of the required proovable `Val`

### Root Hash Calculation of the minimal subtree

The root hash of the minimal subtree use similar function as in the IntegriTree, see the `aiken`'s implementation below:

``` gleam
pub fn root_hash(root: Proof) -> Hash {
  when root is {
    NodeHash { hash } -> hash
    HashNode { hash, left, right } ->
      combine_three_hashes(hash, root_hash(left), root_hash(right))
  }
}
```

> ## Membership and Non-Membership Proofs

Reviewer: @ilap very important to maintain standard document sections.

IntegriTree provides proofs for both membership and non-membership of elements in the tree:

- **Membership**: If an `x` element is in the tree, there exists at least one node where either `xa = x` or `xb = x`.
- **Non-Membership**: If an element is not in the tree, there exists a node where `xa < x < xb`.

> ## Append and Delete Proofs

Reviewer 4: ```suggestion
### Append and Delete Proofs
```

Reviewer: @ilap very important to maintain standard document sections.

Proofs are available for appending new elements to the tree and deleting existing elements:

- **Append Proof**: If an element is not in the tree, the proof contains the non-membership (update node's) value `Vu`, the value `Vp` of the node will be the parent of the new appendable node of the new element, and the minimal subtree of the two `Val`'s nodes.
- **Delete Proof**: If an element is in the tree, the proof contains the two membership (update nodes') values the `Vu1` and `Vu2` (the `(xα, x)` and `(x, xβ)`),  and the value `Vp` the parent node of the last (in level-order) node in the tree, and the minimal subtree create from the nodes of the three values.

### Val Types of the Proof

There are two types of Vals are used with the Proof:

- **Updateable Node's (`Nu`'s)**: Nodes' val that are already part of the tree but may have their values updated.
- **Parent Node's (`Np`'s)**: Nodes' val where a new leaf will be appended to or removed from.

> ## Completeness Enforcement

Reviewer 4: ```suggestion
### Completeness Enforcement
```

Reviewer: @ilap very important to maintain standard document sections.

The tree must always maintain its completeness which enforced by the sequential index of nodes in the tree.

> ## Technical briefs of the IntegriTree

Reviewer 4: ```suggestion
### Technical briefs of the IntegriTree
```

---
CIP: XXXX
Title: Efficient Proofs for Dynamic Sets on Cardano
Status: Proposed
Category: Tools
Authors:
    - Pal Dorogi <pal.dorogi@gmail.com>

> Implementors: []

Reviewer: fixed with https://github.com/cardano-foundation/CIPs/pull/816/commits/544a0cbaededa41d514026c4eaec1eb451eee807 (for proper introduction at tonight's CIP meeting) - likely result of GitHub vs. local branch merge (cc @ilap).
